### PR TITLE
Add vector actions to IMPALA

### PR DESCRIPTION
**Pull request by stefanpantic**
## What do these changes do?
* Abstracts action space type in IMPALA and allows for vector actions.

### CODE REVIEW

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 1]
```diff
@@ -216,9 +216,8 @@ def __init__(self,
         # Setup the policy
         dist_class, cat_logit_dim = ModelCatalog.get_action_dist(
             action_space, self.config["model"],
-            dist_type=self.config["dist_type"])
-        if self._is_discrete:
-            logit_dim = cat_logit_dim
+            dist_type=self.config['dist_type'])
+        logit_dim = cat_logit_dim or logit_dim
```
Comment (bjg2) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: This should stay behind `if self._is_discrete:`, or to be guarded differently. Point is, we don't want to take logit_dim from multicategorical distribution but from hidden observation shape.

File: python/ray/rllib/agents/impala/multi_vtrace.py [hunk 2]
```diff
@@ -129,6 +129,7 @@ def from_logits(behaviour_policy, target_policy, actions,
             clip_rho_threshold=clip_rho_threshold,
             clip_pg_rho_threshold=clip_pg_rho_threshold)
 
+        # noinspection PyProtectedMember
```
Comment (bjg2) [inline python/ray/rllib/agents/impala/multi_vtrace.py]: Let's remove these everywhere (shouldn't be a part of the pull request as they don't use them)

File: .pep8 [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[pep8]
```
Comment (bjg2) [inline .pep8]: Let's remove this one (shouldn't be a part of the pull request as they don't use them)

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 4]
```diff
@@ -274,52 +283,33 @@ def make_time_major(tensor, drop_last=False):
...
+            rewards=make_time_major(rewards, drop_last=True),
+            values=make_time_major(values, drop_last=True),
+            bootstrap_value=make_time_major(values)[-1],
+            valid_mask=make_time_major(mask, drop_last=True),
+            vf_loss_coeff=self.config["vf_loss_coeff"],
+            entropy_coeff=self.config["entropy_coeff"],
+            clip_rho_threshold=self.config["vtrace_clip_rho_threshold"],
+            clip_pg_rho_threshold=self.config["vtrace_clip_pg_rho_threshold"])
+
+        model_dist = dist_class(dist_inputs)
```
Comment (bjg2) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py]: This is not ok - in the vanilla ray they were Categorical even though dist_class was not (it was DiagGaussian). So revert to either Categorical or MultiCategorical, behind discrete if.

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 5]
```diff
@@ -274,52 +283,33 @@ def make_time_major(tensor, drop_last=False):
...
+        self.loss = MultiVTraceLoss(
+            actions=make_time_major(loss_actions, drop_last=True),
+            actions_logp=make_time_major(action_dist.logp(actions),
+                                         drop_last=True),
+            actions_entropy=make_time_major(action_dist.entropy(),
+                                            drop_last=True),
+            dones=make_time_major(dones, drop_last=True),
+            behaviour_logits=make_time_major(
+                unpacked_behaviour_logits, drop_last=True),
+            target_logits=make_time_major(unpacked_outputs, drop_last=True),
```
Comment (bjg2) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py:303]: target_logits=make_time_major(dist_inputs, drop_last=True),

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 6]
```diff
@@ -235,15 +234,25 @@ def __init__(self,
             seq_lens=existing_seq_lens)
         unpacked_outputs = tf.split(
             self.model.outputs, self.output_hidden_shape, axis=1)
-        if self._is_discrete:
-            action_dist = dist_class(self.model.outputs)
-        else:
-            action_dist = dist_class(unpacked_outputs)
+        dist_inputs = self.model.outputs if self._is_discrete \
```
Comment (bjg2) [inline python/ray/rllib/agents/impala/vtrace_policy_graph.py:240]: I think that dist_inputs could just be unpacked_outputs? Not sure, think about it.

But whatever you do with model.outputs / unpacked outputs you should mirror it to behavior_logits / unpacked_behavior_logits - that is not the case currently.

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py (hunk 5 above)
Comment (stefanpantic) [reply python/ray/rllib/agents/impala/vtrace_policy_graph.py:303]: Our VTrace requires this to be a list, unstacked_outputs is basically [dist_inputs] in the discrete case

File: python/ray/rllib/models/catalog.py [hunk 7]
```diff
@@ -116,12 +116,15 @@ def get_action_dist(action_space, config, dist_type=None):
...
                         "clip_actions agent option instead.")
                 return dist, action_space.shape[0] * 2
-            elif dist_type == "categorical":
+            elif dist_type == "multi_categorical":
                 return MultiCategorical, None
             elif dist_type == "deterministic":
                 return Deterministic, action_space.shape[0]
         elif isinstance(action_space, gym.spaces.Discrete):
-            return Categorical, action_space.n
+            if dist_type == "multi_categorical":
```
Comment (bjg2) [inline python/ray/rllib/models/catalog.py]: What is the example of the multicategorical discrete distribution

File: python/ray/rllib/models/action_dist.py [hunk 8]
```diff
@@ -102,12 +102,12 @@ def sample(self):
...
 
+    # noinspection PyMissingConstructor
     def __init__(self, inputs):
-        self.cats = []
-        for input in inputs:
-            self.cats.append(Categorical(input))
+        self.cats = [Categorical(input_) for input_ in inputs]
 
     def logp(self, actions):
+        actions = tf.unstack(actions, axis=1)
```
Comment (bjg2) [inline python/ray/rllib/models/action_dist.py]: Hmmm, I'm not sure. Think about whether it is more general API that receives a list of tensors or tensor of this shape, regardless of our current usage (would it be more general / better code if the unstack would be executed in the caller).

**Review by bjg2 — COMMENTED** [review]
this is my review :)

File: python/ray/rllib/models/action_dist.py (hunk 8 above)
Comment (stefanpantic) [reply python/ray/rllib/models/action_dist.py]: Paging @nikola-j

**Review by nikola-j — APPROVED** [review]

**Review by bjg2 — APPROVED** [review]
hvala poy

**Pull request merged**